# 1. Import Libraries

In [36]:
#Import libraries
import pandas as pd
import numpy as np
import os


# 2. Import Data

In [37]:
path = r"C:\Users\marce\Documents\Apartment rental offers in Germany"

In [38]:
#Import 'immo_data' data set
df=pd.read_csv(os.path.join (path, '02 Data', 'Original Data', 'immo_data.csv'), index_col =False)


In [39]:
df.shape

(268850, 49)

In [40]:
df.head()

regio1  serviceCharge                     heatingType  \
0  Nordrhein_Westfalen         245.00                 central_heating   
1      Rheinland_Pfalz         134.00  self_contained_central_heating   
2              Sachsen         255.00                   floor_heating   
3              Sachsen          58.15                district_heating   
4               Bremen         138.00  self_contained_central_heating   

  telekomTvOffer  telekomHybridUploadSpeed  newlyConst  balcony  picturecount  \
0  ONE_YEAR_FREE                       NaN       False    False             6   
1  ONE_YEAR_FREE                       NaN       False     True             8   
2  ONE_YEAR_FREE                      10.0        True     True             8   
3  ONE_YEAR_FREE                       NaN       False     True             9   
4            NaN                       NaN       False     True            19   

   pricetrend  telekomUploadSpeed  ...               regio2  \
0        4.62                10.0  ...             Dortmund   
1        3.47                10.0  ...    Rhein_Pfalz_Kreis   
2        2.72                 2.4  ...              Dresden   
3        1.53                40.0  ...  Mittelsachsen_Kreis   
4        2.46                 NaN  ...               Bremen   

                       regio3  \
0                     Schüren   
1              Böhl_Iggelheim   
2  Äußere_Neustadt_Antonstadt   
3                    Freiberg   
4           Neu_Schwachhausen   

                                         description  \
0  Die ebenerdig zu erreichende Erdgeschosswohnun...   
1  Alles neu macht der Mai – so kann es auch für ...   
2  Der Neubau entsteht im Herzen der Dresdner Neu...   
3  Abseits von Lärm und Abgasen in Ihre neue Wohn...   
4  Es handelt sich hier um ein saniertes Mehrfami...   

                                          facilities heatingCosts  \
0  Die Wohnung ist mit Laminat ausgelegt. Das Bad...          NaN   
1                                                NaN          NaN   
2  * 9 m² Balkon\n* Bad mit bodengleicher Dusche,...          NaN   
3                                                NaN        87.23   
4  Diese Wohnung wurde neu saniert und ist wie fo...          NaN   

   energyEfficiencyClass lastRefurbish  electricityBasePrice  \
0                    NaN           NaN                   NaN   
1                    NaN        2019.0                   NaN   
2                    NaN           NaN                   NaN   
3                    NaN           NaN                   NaN   
4                    NaN           NaN                   NaN   

   electricityKwhPrice   date  
0                  NaN  May19  
1                  NaN  May19  
2                  NaN  Oct19  
3                  NaN  May19  
4                  NaN  Feb20  

[5 rows x 49 columns]

# 3. Data Quality Check

In [41]:
# Check data set info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 268850 entries, 0 to 268849
Data columns (total 49 columns):
 #   Column                    Non-Null Count   Dtype  
---  ------                    --------------   -----  
 0   regio1                    268850 non-null  object 
 1   serviceCharge             261941 non-null  float64
 2   heatingType               223994 non-null  object 
 3   telekomTvOffer            236231 non-null  object 
 4   telekomHybridUploadSpeed  45020 non-null   float64
 5   newlyConst                268850 non-null  bool   
 6   balcony                   268850 non-null  bool   
 7   picturecount              268850 non-null  int64  
 8   pricetrend                267018 non-null  float64
 9   telekomUploadSpeed        235492 non-null  float64
 10  totalRent                 228333 non-null  float64
 11  yearConstructed           211805 non-null  float64
 12  scoutId                   268850 non-null  int64  
 13  noParkSpaces              93052 non-null   f

In [42]:
# Descriptive Statistics
df.describe()

serviceCharge  telekomHybridUploadSpeed   picturecount     pricetrend  \
count  261941.000000                   45020.0  268850.000000  267018.000000   
mean      151.206113                      10.0       9.791958       3.389001   
std       308.295790                       0.0       6.408399       1.964874   
min         0.000000                      10.0       0.000000     -12.330000   
25%        95.000000                      10.0       6.000000       2.000000   
50%       135.000000                      10.0       9.000000       3.390000   
75%       190.000000                      10.0      13.000000       4.570000   
max    146118.000000                      10.0     121.000000      14.920000   

       telekomUploadSpeed     totalRent  yearConstructed       scoutId  \
count       235492.000000  2.283330e+05    211805.000000  2.688500e+05   
mean            28.804928  9.013315e+02      1966.400590  1.069697e+08   
std             16.337151  3.323833e+04        46.992207  1.250093e+07   
min              1.000000  0.000000e+00      1000.000000  2.887174e+07   
25%             10.000000  4.698000e+02      1950.000000  1.066910e+08   
50%             40.000000  6.500000e+02      1973.000000  1.111584e+08   
75%             40.000000  9.850000e+02      1996.000000  1.137688e+08   
max            100.000000  1.575154e+07      2090.000000  1.157117e+08   

       noParkSpaces  yearConstructedRange  ...        noRooms    thermalChar  \
count  93052.000000         211805.000000  ...  268850.000000  162344.000000   
mean       1.327634              3.714544  ...       2.641261     114.749533   
std        8.361403              2.738134  ...       2.633440      61.653663   
min        0.000000              1.000000  ...       1.000000       0.100000   
25%        1.000000              1.000000  ...       2.000000      79.000000   
50%        1.000000              3.000000  ...       3.000000     107.000000   
75%        1.000000              5.000000  ...       3.000000     140.300000   
max     2241.000000              9.000000  ...     999.990000    1996.000000   

               floor  numberOfFloors   noRoomsRange  livingSpaceRange  \
count  217541.000000   171118.000000  268850.000000     268850.000000   
mean        2.122405        3.572319       2.571542          3.070790   
std         3.634934        6.375496       0.937594          1.407127   
min        -1.000000        0.000000       1.000000          1.000000   
25%         1.000000        2.000000       2.000000          2.000000   
50%         2.000000        3.000000       3.000000          3.000000   
75%         3.000000        4.000000       3.000000          4.000000   
max       999.000000      999.000000       5.000000          7.000000   

       heatingCosts  lastRefurbish  electricityBasePrice  electricityKwhPrice  
count  85518.000000   80711.000000          46846.000000         46846.000000  
mean      76.990866    2013.904536             89.113612             0.199769  
std      147.716278      10.963125              5.395805             0.009667  
min        0.000000    1015.000000             71.430000             0.170500  
25%       54.000000    2012.000000             90.760000             0.191500  
50%       70.000000    2017.000000             90.760000             0.198500  
75%       90.000000    2019.000000             90.760000             0.205500  
max    12613.000000    2919.000000             90.760000             0.227600  

[8 rows x 24 columns]

Note: There are many erroneous data such as service Charge (146118), picturecount(121), noParkSpaces (2241), noRooms-999, floor 999, numberOfFloors 999, lastRefurbish 2919

In [43]:
# Check for mixed data  types

for col in df.columns.tolist():
  weird = (df[[col]].applymap(type) != df[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df[weird]) > 0:
    print (col)

heatingType
telekomTvOffer
firingTypes
houseNumber
condition
interiorQual
petsAllowed
streetPlain
typeOfFlat
description
facilities
energyEfficiencyClass


# 3.1 Missing Value

In [44]:
# Count the missing values in each column:
missing_values =df.isnull().sum()

In [45]:
# Calculate the total number of rows in the DataFrame
total_cells = df.shape[0]

# Calculate the percentage of missing values
percent_missing = (missing_values / total_cells) * 100

# Round the decimal to 2 decimal places
percent_missing = percent_missing.round(2)

# Convert the float type to string
percent_missing = percent_missing.astype(str)

# Add the '%' symbol to each value
percent_missing = percent_missing + '%'

# Combine the two Series into a new DataFrame
missing_value_table = pd.concat([missing_values, percent_missing], axis=1)

# Rename the columns of the new DataFrame
missing_value_table.columns = ['Missing Values', 'Percent Missing']

# Sort the resulting DataFrame based on the 'Percent Missing' column in ascending order
missing_value_table.sort_values('Percent Missing', ascending=True, inplace=True)

# Print the resulting table
print(missing_value_table)

                          Missing Values Percent Missing
regio1                                 0            0.0%
regio3                                 0            0.0%
regio2                                 0            0.0%
livingSpaceRange                       0            0.0%
garden                                 0            0.0%
noRoomsRange                           0            0.0%
noRooms                                0            0.0%
geo_plz                                0            0.0%
baseRentRange                          0            0.0%
lift                                   0            0.0%
street                                 0            0.0%
geo_krs                                0            0.0%
livingSpace                            0            0.0%
baseRent                               0            0.0%
cellar                                 0            0.0%
geo_bln                                0            0.0%
date                           

# 3.2 Duplicates

In [46]:
# Remove duplicates
df.drop_duplicates(inplace=True)

In [73]:
# Data contains no duplicates

# 3.3 Droping unnecessary columns

In [47]:
# creating a new dataframe and dropping the columns with more than 30% missing value
threshold = 0.3
missing_value_table['Percent Missing'] = missing_value_table['Percent Missing'].str.strip('%').astype(float)
missing_value_columns = missing_value_table[missing_value_table['Percent Missing'] >= threshold*100].index
df_new = df.drop(missing_value_columns, axis=1)

In [48]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 268850 entries, 0 to 268849
Data columns (total 38 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   regio1                268850 non-null  object 
 1   serviceCharge         261941 non-null  float64
 2   heatingType           223994 non-null  object 
 3   telekomTvOffer        236231 non-null  object 
 4   newlyConst            268850 non-null  bool   
 5   balcony               268850 non-null  bool   
 6   picturecount          268850 non-null  int64  
 7   pricetrend            267018 non-null  float64
 8   telekomUploadSpeed    235492 non-null  float64
 9   totalRent             228333 non-null  float64
 10  yearConstructed       211805 non-null  float64
 11  scoutId               268850 non-null  int64  
 12  firingTypes           211886 non-null  object 
 13  hasKitchen            268850 non-null  bool   
 14  geo_bln               268850 non-null  object 
 15  

In [49]:
# dropping unwanted columns
df_new=df_new.drop(['scoutId', 'yearConstructedRange', 'baseRentRange','streetPlain','noRoomsRange', 'livingSpaceRange', 'description',],axis=1)

In [50]:
df_new.head()

regio1  serviceCharge                     heatingType  \
0  Nordrhein_Westfalen         245.00                 central_heating   
1      Rheinland_Pfalz         134.00  self_contained_central_heating   
2              Sachsen         255.00                   floor_heating   
3              Sachsen          58.15                district_heating   
4               Bremen         138.00  self_contained_central_heating   

  telekomTvOffer  newlyConst  balcony  picturecount  pricetrend  \
0  ONE_YEAR_FREE       False    False             6        4.62   
1  ONE_YEAR_FREE       False     True             8        3.47   
2  ONE_YEAR_FREE        True     True             8        2.72   
3  ONE_YEAR_FREE       False     True             9        1.53   
4            NaN       False     True            19        2.46   

   telekomUploadSpeed  totalRent  ...   lift    typeOfFlat  geo_plz noRooms  \
0                10.0      840.0  ...  False  ground_floor    44269     4.0   
1                10.0        NaN  ...  False  ground_floor    67459     3.0   
2                 2.4     1300.0  ...   True     apartment     1097     3.0   
3                40.0        NaN  ...  False         other     9599     3.0   
4                 NaN      903.0  ...  False     apartment    28213     3.0   

   floor  garden               regio2                      regio3  \
0    1.0    True             Dortmund                     Schüren   
1    NaN   False    Rhein_Pfalz_Kreis              Böhl_Iggelheim   
2    3.0   False              Dresden  Äußere_Neustadt_Antonstadt   
3    3.0   False  Mittelsachsen_Kreis                    Freiberg   
4    1.0   False               Bremen           Neu_Schwachhausen   

                                          facilities   date  
0  Die Wohnung ist mit Laminat ausgelegt. Das Bad...  May19  
1                                                NaN  May19  
2  * 9 m² Balkon\n* Bad mit bodengleicher Dusche,...  Oct19  
3                                                NaN  May19  
4  Diese Wohnung wurde neu saniert und ist wie fo...  Feb20  

[5 rows x 31 columns]

In [51]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 268850 entries, 0 to 268849
Data columns (total 31 columns):
 #   Column              Non-Null Count   Dtype  
---  ------              --------------   -----  
 0   regio1              268850 non-null  object 
 1   serviceCharge       261941 non-null  float64
 2   heatingType         223994 non-null  object 
 3   telekomTvOffer      236231 non-null  object 
 4   newlyConst          268850 non-null  bool   
 5   balcony             268850 non-null  bool   
 6   picturecount        268850 non-null  int64  
 7   pricetrend          267018 non-null  float64
 8   telekomUploadSpeed  235492 non-null  float64
 9   totalRent           228333 non-null  float64
 10  yearConstructed     211805 non-null  float64
 11  firingTypes         211886 non-null  object 
 12  hasKitchen          268850 non-null  bool   
 13  geo_bln             268850 non-null  object 
 14  cellar              268850 non-null  bool   
 15  baseRent            268850 non-nul

In [52]:
# Remove more unnecessary columns

columns_drop2= ['date','regio2', 'regio3','street','firingTypes', 'telekomUploadSpeed','geo_bln','facilities', 'houseNumber']
df_new.drop(columns_drop2, axis=1, inplace=True)

In [53]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 268850 entries, 0 to 268849
Data columns (total 22 columns):
 #   Column           Non-Null Count   Dtype  
---  ------           --------------   -----  
 0   regio1           268850 non-null  object 
 1   serviceCharge    261941 non-null  float64
 2   heatingType      223994 non-null  object 
 3   telekomTvOffer   236231 non-null  object 
 4   newlyConst       268850 non-null  bool   
 5   balcony          268850 non-null  bool   
 6   picturecount     268850 non-null  int64  
 7   pricetrend       267018 non-null  float64
 8   totalRent        228333 non-null  float64
 9   yearConstructed  211805 non-null  float64
 10  hasKitchen       268850 non-null  bool   
 11  cellar           268850 non-null  bool   
 12  baseRent         268850 non-null  float64
 13  livingSpace      268850 non-null  float64
 14  geo_krs          268850 non-null  object 
 15  condition        200361 non-null  object 
 16  lift             268850 non-null  bool

In [54]:
df_new.head()

regio1  serviceCharge                     heatingType  \
0  Nordrhein_Westfalen         245.00                 central_heating   
1      Rheinland_Pfalz         134.00  self_contained_central_heating   
2              Sachsen         255.00                   floor_heating   
3              Sachsen          58.15                district_heating   
4               Bremen         138.00  self_contained_central_heating   

  telekomTvOffer  newlyConst  balcony  picturecount  pricetrend  totalRent  \
0  ONE_YEAR_FREE       False    False             6        4.62      840.0   
1  ONE_YEAR_FREE       False     True             8        3.47        NaN   
2  ONE_YEAR_FREE        True     True             8        2.72     1300.0   
3  ONE_YEAR_FREE       False     True             9        1.53        NaN   
4            NaN       False     True            19        2.46      903.0   

   yearConstructed  ...  baseRent  livingSpace              geo_krs  \
0           1965.0  ...     595.0        86.00             Dortmund   
1           1871.0  ...     800.0        89.00    Rhein_Pfalz_Kreis   
2           2019.0  ...     965.0        83.80              Dresden   
3           1964.0  ...     343.0        58.15  Mittelsachsen_Kreis   
4           1950.0  ...     765.0        84.97               Bremen   

        condition   lift    typeOfFlat  geo_plz noRooms  floor  garden  
0       well_kept  False  ground_floor    44269     4.0    1.0    True  
1     refurbished  False  ground_floor    67459     3.0    NaN   False  
2  first_time_use   True     apartment     1097     3.0    3.0   False  
3             NaN  False         other     9599     3.0    3.0   False  
4     refurbished  False     apartment    28213     3.0    1.0   False  

[5 rows x 22 columns]

# 3.4 Renaming columns

In [55]:
#Rename geo columns
df_new.rename(columns={'regio1':'State', 'geo_plz':'Pincode', 'geo_krs':'City' }, inplace=True)

In [56]:
#Rename house columns
df_new.rename(columns={'heatingType':'Heating_type','telekomTvOffer':'TV_offer', 'newlyConst':'Newly_built', 'balcony':'Balcony', 'yearConstructed':'Year_constructed', 'hasKitchen':'Kitchen', 'cellar':'Cellar', 'livingSpace':'Area(m2)','condition':'House_condition','lift':'Lift','typeOfFlat':'Flat_type', 'floor':'Floor','noRoomsRange':'Rooms', 'garden':'Garden' }, inplace=True)

In [57]:
#Rename Price columns
df_new.rename(columns={'serviceCharge':'Maintenance_fee','pricetrend':'Price_trend', 'baseRent':'Base_rent', 'picturecount':'Picture_count' }, inplace=True)


In [58]:
#Rename 'totalRent' and 'noRooms' columns
df_new.rename(columns={'totalRent':'Total_rent', 'noRooms':'Rooms_number' }, inplace=True)

In [59]:
df_new.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 268850 entries, 0 to 268849
Data columns (total 22 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   State             268850 non-null  object 
 1   Maintenance_fee   261941 non-null  float64
 2   Heating_type      223994 non-null  object 
 3   TV_offer          236231 non-null  object 
 4   Newly_built       268850 non-null  bool   
 5   Balcony           268850 non-null  bool   
 6   Picture_count     268850 non-null  int64  
 7   Price_trend       267018 non-null  float64
 8   Total_rent        228333 non-null  float64
 9   Year_constructed  211805 non-null  float64
 10  Kitchen           268850 non-null  bool   
 11  Cellar            268850 non-null  bool   
 12  Base_rent         268850 non-null  float64
 13  Area(m2)          268850 non-null  float64
 14  City              268850 non-null  object 
 15  House_condition   200361 non-null  object 
 16  Lift              26

# 3.5 Replace missing values

In [60]:
# Replacing missing values with NaN
df_new.fillna('NaN', inplace=True)

# 3.6 Convert Datatypes

In [71]:
#Convert 'Year_cosntruted' Datatype to Int

df_new['Year_constructed']=df_new['Year_constructed'].astype(float).astype('Int64')


In [62]:
#Convert 'Rooms_number' Datatype to Int

df_new['Rooms_number']=df_new['Rooms_number'].astype(int)



In [63]:
# checking the values for year column
df_new['Year_constructed'].value_counts()

2019    10959
1900    10356
2018     8759
1995     4387
1996     4339
        ...  
1586        1
1027        1
1787        1
1627        1
1472        1
Name: Year_constructed, Length: 465, dtype: Int64

In [64]:
# checking the values for year column
df_new['Year_constructed'].value_counts()

2019    10959
1900    10356
2018     8759
1995     4387
1996     4339
        ...  
1586        1
1027        1
1787        1
1627        1
1472        1
Name: Year_constructed, Length: 465, dtype: Int64

In [65]:
#Removing outlier 2090
df_new1 = df_new[df_new['Year_constructed'] != 2090]

In [66]:
# keeping the data only with the year<2023 and >1800
df_new2 = df_new1[(df_new1['Year_constructed']<2023) & (df_new1['Year_constructed']>1800)]

In [68]:
df_new2.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 210771 entries, 0 to 268849
Data columns (total 22 columns):
 #   Column            Non-Null Count   Dtype  
---  ------            --------------   -----  
 0   State             210771 non-null  object 
 1   Maintenance_fee   210771 non-null  object 
 2   Heating_type      210771 non-null  object 
 3   TV_offer          210771 non-null  object 
 4   Newly_built       210771 non-null  bool   
 5   Balcony           210771 non-null  bool   
 6   Picture_count     210771 non-null  int64  
 7   Price_trend       210771 non-null  object 
 8   Total_rent        210771 non-null  object 
 9   Year_constructed  210771 non-null  Int64  
 10  Kitchen           210771 non-null  bool   
 11  Cellar            210771 non-null  bool   
 12  Base_rent         210771 non-null  float64
 13  Area(m2)          210771 non-null  float64
 14  City              210771 non-null  object 
 15  House_condition   210771 non-null  object 
 16  Lift              21

In [69]:
df_new2.head()

State Maintenance_fee                    Heating_type  \
0  Nordrhein_Westfalen           245.0                 central_heating   
1      Rheinland_Pfalz           134.0  self_contained_central_heating   
2              Sachsen           255.0                   floor_heating   
3              Sachsen           58.15                district_heating   
4               Bremen           138.0  self_contained_central_heating   

        TV_offer  Newly_built  Balcony  Picture_count Price_trend Total_rent  \
0  ONE_YEAR_FREE        False    False              6        4.62      840.0   
1  ONE_YEAR_FREE        False     True              8        3.47        NaN   
2  ONE_YEAR_FREE         True     True              8        2.72     1300.0   
3  ONE_YEAR_FREE        False     True              9        1.53        NaN   
4            NaN        False     True             19        2.46      903.0   

   Year_constructed  ...  Base_rent  Area(m2)                 City  \
0              1965  ...      595.0     86.00             Dortmund   
1              1871  ...      800.0     89.00    Rhein_Pfalz_Kreis   
2              2019  ...      965.0     83.80              Dresden   
3              1964  ...      343.0     58.15  Mittelsachsen_Kreis   
4              1950  ...      765.0     84.97               Bremen   

   House_condition   Lift     Flat_type  Pincode Rooms_number  Floor  Garden  
0        well_kept  False  ground_floor    44269            4    1.0    True  
1      refurbished  False  ground_floor    67459            3    NaN   False  
2   first_time_use   True     apartment     1097            3    3.0   False  
3              NaN  False         other     9599            3    3.0   False  
4      refurbished  False     apartment    28213            3    1.0   False  

[5 rows x 22 columns]

In [72]:
# Exporting the data 
df_new2.to_csv(os.path.join(path,'02 Data','Prepared Data','immodata_cleaned.csv'))